diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -543,29 +552,117 @@ bool DifferentialFunctionComparator::mayIgnore(const User *Inst) const {
                 // loaded back.
                 auto IntTySrc = dyn_cast<IntegerType>(SrcTy);
                 auto IntTyDest = dyn_cast<IntegerType>(DestTy);
-                if (IntTySrc->getBitWidth() > IntTyDest->getBitWidth()) {
-                    return false;
+                if (IntTySrc->getBitWidth() <= IntTyDest->getBitWidth()) {
+                    // Look for arithmetic operations in the uses of the cast
+                    // and in the uses of all values that are generated by
+                    // further casting.
+                    std::vector<const User *> UserStack;
+                    UserStack.push_back(Inst);
+                    while (!UserStack.empty()) {
+                        const User *U = UserStack.back();
+                        UserStack.pop_back();
+                        if (isa<BinaryOperator>(U)) {
+                            return false;
+                        } else if (isa<CastInst>(U)) {
+                            for (const User *UU : U->users())
+                                UserStack.push_back(UU);
+                        }
+                    }
+                    canBeIgnored = true;
+                    replacement = Inst->getOperand(0);
                 }
-                // Look for arithmetic operations in the uses of the cast and
-                // in the uses of all values that are generated by further
-                // casting.
-                std::vector<const User *> UserStack;
-                UserStack.push_back(Inst);
-                while (!UserStack.empty()) {
-                    const User *U = UserStack.back();
-                    UserStack.pop_back();
-                    if (isa<BinaryOperator>(U)) {
-                        return false;
-                    } else if (isa<CastInst>(U)) {
-                        for (const User *UU : U->users())
-                            UserStack.push_back(UU);
+            }
+        } else if (auto Load = dyn_cast<LoadInst>(Inst)) {
+            return mayIgnoreLoad(Load, addToIgnored);
+        }
+    }
+
+    if (canBeIgnored && addToIgnored) {
+        ignoredInstructions.insert({Inst, replacement});
+    }
+    return canBeIgnored;
+}
+
+/// Check whether the given instruction is a repetitive variant of a previous
+/// load with no store instructions in between. Will return the first load
+/// if the replacement pointer is given.
+bool DifferentialFunctionComparator::mayIgnoreLoad(const LoadInst *Load,
+                                                   bool addToIgnored) const {
+    if (auto BB = Load->getParent()) {
+        // Check if all predecessor blocks have a path to a load
+        // from the same pointer with no stores to the same pointer
```

Comment:
In fact, you're not checking for stores to the *same* pointer but to *any* pointer since checking the same pointer is not sound because of pointer aliasing. The comment should state that.